src/main/java/io/nats/client/impl/NatsConnection.java

```diff
@@ -584,36 +584,25 @@ void checkVersionRequirements() throws IOException {
     }
 
     void upgradeToSecureIfNeeded(NatsUri nuri) throws IOException {
-        Options clientOptions = getOptions();
-        if (clientOptions.isTlsFirst()) {
-            this.dataPort.upgradeToSecure();
+        if (options.isTlsFirst()) {
+            dataPort.upgradeToSecure();
         }
         else {
             ServerInfo serverInfo = getInfo();
-            boolean before2_9_19 = serverInfo.isOlderThanVersion("2.9.19");
-
-            boolean isTLSRequired = clientOptions.isTLSRequired();
-            boolean upgradeRequired = isTLSRequired;
-            if (isTLSRequired && nuri.isWebsocket()) {
-                // We are already communicating over "https" websocket, so
-                // do NOT try to upgrade to secure.
-                if (before2_9_19) {
-                    isTLSRequired = false;
+            if (options.isTLSRequired()) {
+                if (!nuri.isWebsocket()) {
+                    // When already communicating over "https" websocket, do NOT try to upgrade to secure.
+                    if (!serverInfo.isTLSRequired() && !serverInfo.isTLSAvailable()) {
+                        throw new IOException("SSL connection wanted by client.");
+                    }
+                    dataPort.upgradeToSecure();
                 }
-                upgradeRequired = false;
-            }
-            if (isTLSRequired && !serverInfo.isTLSRequired()) {
-                throw new IOException("SSL connection wanted by client.");
             }
-            else if (!isTLSRequired && serverInfo.isTLSRequired()) {
+            else if (serverInfo.isTLSRequired()) {
                 throw new IOException("SSL required by server.");
             }
-            if (upgradeRequired) {
-                this.dataPort.upgradeToSecure();
-            }
         }
     }
-

```

Comment:
This whole block is simplified.
1. getOptions returns the options variable, just use the variable directly. There is no lazy loading or anything.
2. No need to check 2.9.19 because websocket never does upgrade, it was just extraneous